subprojects/base-services/src/main/java/org/gradle/internal/classloader/CachingClassLoader.java

```diff
@@ -20,13 +20,34 @@
 
 import java.io.Closeable;
 import java.io.IOException;
+import java.lang.reflect.Method;
+import java.lang.reflect.InvocationTargetException;
 import java.util.concurrent.ConcurrentMap;
 
 public class CachingClassLoader extends ClassLoader implements ClassLoaderHierarchy, Closeable {
     private static final Object MISSING_CLASS = new Object();
     private final ConcurrentMap<String, Object> loadedClasses = new MapMaker().weakValues().makeMap();
     private final ClassLoader parent;
 
+    static {
+        /*
+         * This classloader is thread-safe and ClassLoader is parallel capable,
+         * so register as such to reduce contention when running multithreaded builds.
+         * We do so through relfection since Gradle should print error messages when
+         * run with older JRE versions
+        */
+        try {
+            Method m = ClassLoader.class.getMethod("registerAsParallelCapable");
```

Comment:
@jochenberger Keep in mind that the `setAccessible(true)` approach might stop working in Java 9.

See [setAccessible(boolean)](http://download.java.net/java/jdk9/docs/api/java/lang/reflect/AccessibleObject.html#setAccessible-boolean-):
> This method cannot be used to enable access to an object that is a member of a class in a different module to the caller and where the class is in a package that is not exported to the caller's module. Additionally, if the member is non-public or its declaring class is non-public, then this method can only be used to enable access if the package is open to at least the caller's module. 

I'm not sure if this applies since the `CachingClassLoader` would be in the caller's module while the declaring class, i.e. `java.lang.ClassLoader`, wouldn't. But I guess it applies because `registerAsParallelCapable()` is a static method.